ooniprobe/Test/Test/AbstractTest.mm

```diff
@@ -58,16 +58,24 @@ -(void)prepareRun{
 -(void)runTest{
     if(self.annotation)
         [self.settings.annotations setObject:@"ooni-run" forKey:@"origin"];
-    NSDictionary *settings = [self.settings getSettingsDictionary];
     self.backgroundTask = [[UIApplication sharedApplication] beginBackgroundTaskWithExpirationHandler:^{
         [[UIApplication sharedApplication] endBackgroundTask:self.backgroundTask];
         self.backgroundTask = UIBackgroundTaskInvalid;
     }];
     dispatch_async(_serialQueue, ^{
-        MKAsyncTask *task = [MKAsyncTask start:settings];
-        while (![task done]){
+        NSError *error;
+        id<ExperimentTask> task = [Engine startExperimentTaskWithSettings:[self.settings toExperimentSettings] error:&error];
+        if (error != nil){
+            self.result.failure_msg = error.description;
+            [self.result save];
```

Comment:
Why do you save here?

Reply:
Because it's a DB object and I just edited some row/column

Reply:
gotcha